File: scripts/RegretFingerprint.java
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import java.math.BigInteger;

/**
 * RegretFingerprint — port of scripts/fingerprint.js to Java.
 *
 * Algorithm IDENTICAL to fingerprint.js / fingerprint.py / fingerprint.go:
 *   1. cleanInput  = stripFields(normalize(input,  rules), ignoreFields)
 *   2. cleanOutput = stripFields(normalize(output, rules), ignoreFields)
 *   3. combined = stableStringify(cleanInput) + "|" + stableStringify(cleanOutput)
 *   4. hash = SHA-256(combined) as hex
 *   5. num = BigInteger(hex, 16)
 *   6. base36 = num.toString(36)
 *   7. return base36.substring(0, 7)
 *
 * Usage from a generated capture/validate test:
 *   String fp = RegretFingerprint.fingerprint(input, output);
 *
 * The `input` and `output` arguments may be any JSON-serializable Java value:
 *   null, Boolean, Integer, Long, Double, String, List<Object>, Map<String,Object>.
 * Maps MUST use String keys (mirrors JS/Python object semantics).
 *
 * Cross-stack parity verified against scripts/fingerprint.js — same input/output
 * pair produces the same 7-char fingerprint in JS and Java.
 */
public class RegretFingerprint {

    // ─── stableStringify ────────────────────────────────────────────────────
    //
    // Recursive deterministic JSON serialization. Object keys are sorted
    // alphabetically. Mirrors stableStringify() in fingerprint.js exactly.
    //
    // Notable behaviors that MUST match the JS implementation:
    //   - null  -> "null"
    //   - Double.NaN           -> "\"__nan__\""          (issue #322)
    //   - Double.POSITIVE_INFINITY -> "\"__infinity__\""
    //   - Double.NEGATIVE_INFINITY -> "\"__neg_infinity__\""
    //   - String -> JSON.stringify(s) (with surrounding quotes & escaping)
    //   - List   -> "[item0,item1,...]"
    //   - Map    -> "{\"k0\":v0,\"k1\":v1,...}" with sorted keys
    //   - Double 1.0 -> "1" (Java BigInteger toString drops trailing .0 — but we use
    //     String.valueOf which keeps "1.0". This is intentional and MATCHES the JS
    //     behavior of Number.toString().)

    public static String stableStringify(Object obj) {
        return doStableStringify(obj);
    }

    private static String doStableStringify(Object obj) {
        if (obj == null) return "null";

        if (obj instanceof Boolean) return obj.toString();

        if (obj instanceof Number) {
            Number n = (Number) obj;
            // Double / Float path — handle NaN, Infinity, and finite values
            if (n instanceof Double || n instanceof Float) {
                double d = n.doubleValue();
                if (Double.isNaN(d)) return "\"__nan__\"";
                if (d == Double.POSITIVE_INFINITY) return "\"__infinity__\"";
                if (d == Double.NEGATIVE_INFINITY) return "\"__neg_infinity__\"";
                // Whole numbers stored as double: emit as integer (matches JS
                // Number.toString behavior where 1.0 -> "1"). Critical for
                // cross-stack fingerprint parity because JS stores numbers as
                // doubles and serializes 1.0 as "1".
                if (d == Math.rint(d) && !Double.isInfinite(d) && Math.abs(d) < 1e15) {
                    return Long.toString((long) d);
                }
                // Otherwise use Java's default double-to-string. JS uses
                // Number.prototype.toString which produces the same shortest
                // round-trip representation for most values. There may be edge
                // cases where they differ (exponential notation); those should
                // be caught by integration tests.
                return Double.toString(d);
            }
            // Integer / Long / BigInteger / Short / Byte
            return n.toString();
        }

        if (obj instanceof String) {
            return jsonStringEscape((String) obj);
        }

        if (obj instanceof List) {
            List<?> list = (List<?>) obj;
            StringBuilder sb = new StringBuilder("[");
            for (int i = 0; i < list.size(); i++) {
                if (i > 0) sb.append(",");
                sb.append(doStableStringify(list.get(i)));
            }
            sb.append("]");
            return sb.toString();
        }

        if (obj instanceof Map) {
            Map<?, ?> map = (Map<?, ?>) obj;
            // Sort keys alphabetically by their String representation
            TreeMap<String, Object> sorted = new TreeMap<>();
            for (Map.Entry<?, ?> entry : map.entrySet()) {
                sorted.put(String.valueOf(entry.getKey()), entry.getValue());
            }
            StringBuilder sb = new StringBuilder("{");
            boolean first = true;
            for (Map.Entry<String, Object> entry : sorted.entrySet()) {
                if (!first) sb.append(",");
                first = false;
                sb.append(jsonStringEscape(entry.getKey()));
                sb.append(":");
                sb.append(doStableStringify(entry.getValue()));
            }
            sb.append("}");
            return sb.toString();
        }

        // Fallback for unknown types — use String.valueOf. This is a known
        // gap: Java objects without a JSON adapter will produce different
        // output than JS. For PR scope, only JSON-serializable types are
        // supported (String, Number, Boolean, List, Map, null).
        return jsonStringEscape(String.valueOf(obj));
    }

    /**
     * JSON-escape a string and wrap in double quotes.
     * Mirrors JSON.stringify() behavior for strings.
     */
    private static String jsonStringEscape(String s) {
        StringBuilder sb = new StringBuilder(s.length() + 2);
        sb.append('"');
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case '"':  sb.append("\\\""); break;
                case '\\': sb.append("\\\\"); break;
                case '\b': sb.append("\\b"); break;
                case '\f': sb.append("\\f"); break;
                case '\n': sb.append("\\n"); break;
                case '\r': sb.append("\\r"); break;
                case '\t': sb.append("\\t"); break;
                default:
                    if (c < 0x20) {
                        sb.append(String.format("\\u%04x", (int) c));
                    } else {
                        sb.append(c);
                    }
            }
        }
        sb.append('"');
        return sb.toString();
    }

    // ─── fingerprint ────────────────────────────────────────────────────────

    /**
     * Compute the 7-char base36 fingerprint.
     *
     * IDENTICAL algorithm to fingerprint.js:
     *   sha256(stableStringify(input) + "|" + stableStringify(output))
     *   → hex → BigInteger → base36 → first 7 chars
     */
    public static String fingerprint(Object input, Object output) {
        String combined = stableStringify(input) + "|" + stableStringify(output);
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            byte[] hashBytes = md.digest(combined.getBytes(StandardCharsets.UTF_8));
            // Convert to hex string
            StringBuilder hex = new StringBuilder(hashBytes.length * 2);
            for (byte b : hashBytes) {
                hex.append(String.format("%02x", b));
            }
            // BigInt from hex, then base36
            BigInteger num = new BigInteger(hex.toString(), 16);
            String base36 = num.toString(36);
            if (base36.length() < 7) {
                return base36;
            }
            return base36.substring(0, 7);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("SHA-256 not available — JVM misconfiguration", e);
        }
    }

    // ─── .regret file helpers ───────────────────────────────────────────────

    /**
     * Serialize a Java value to a JSON string for the .regret file INPUT/OUTPUT
     * line. Uses stableStringify under the hood so output is deterministic.
     *
     * This is what gets written to the .regret file as `INPUT  <value>` and
     * `OUTPUT <value>`. The validate script will parse this back and re-invoke
     * the function with the parsed input.
     */
    public static String toJsonString(Object obj) {
        return stableStringify(obj);
    }

    /**
     * Construct a complete .regret file content string.
     *
     * Format (mirrors scripts/capture.js / capture.py):
     *   cluster: <id>
     *   version: 1
     *   fingerprint: <fp>
     *   captured: <ISO timestamp>
     *   watches: [<w1>, <w2>, ...]
     *   entry: <entry>
     *   stack: java
     *   fingerprintLevel: <level>
     *   ---
     *   INPUT  <json>
     *   OUTPUT <json>
     *   HASH   <fp>
     */
    public static String buildRegretFile(
            String clusterId,
            String fingerprint,
            String captured,
            String[] watches,
            String entry,
            String fingerprintLevel,
            Object input,
            Object output) {
        StringBuilder sb = new StringBuilder();
        sb.append("cluster: ").append(clusterId).append('\n');
        sb.append("version: 1\n");
        sb.append("fingerprint: ").append(fingerprint).append('\n');
        sb.append("captured: ").append(captured).append('\n');
        sb.append("watches: [").append(String.join(", ", watches)).append("]\n");
        sb.append("entry: ").append(entry).append('\n');
        sb.append("stack: java\n");
        sb.append("fingerprintLevel: ").append(fingerprintLevel).append('\n');
        sb.append("---\n");
        sb.append("INPUT  ").append(toJsonString(input)).append('\n');
        sb.append("OUTPUT ").append(toJsonString(output)).append('\n');
        sb.append("HASH   ").append(fingerprint).append('\n');
        return sb.toString();
    }

    // ─── CLI self-test ──────────────────────────────────────────────────────
    //
    // Run: java RegretFingerprint
    // Expected: prints the fingerprint of (input="hello", output="HELLO")
    // which should match the JS implementation's output for the same input.
    //
    // To verify cross-stack parity:
    //   node -e "import('./scripts/fingerprint.js').then(({fingerprint}) => console.log(fingerprint('hello', 'HELLO')))"
    //   java RegretFingerprint
    //
    // Both should print the same 7-char string.

    public static void main(String[] args) {
        // Test 1: simple string transform
        String fp1 = fingerprint("hello", "HELLO");
        System.out.println("Test 1 (string): " + fp1);
        // Expected: matches `node -e "..."` output for fingerprint('hello', 'HELLO')

        // Test 2: numeric output
        String fp2 = fingerprint(5, 25L);  // 5^2 = 25
        System.out.println("Test 2 (number): " + fp2);

        // Test 3: object output (Map)
        Map<String, Object> input3 = new LinkedHashMap<>();
        input3.put("name", "Alice");
        input3.put("age", 30);
        Map<String, Object> output3 = new LinkedHashMap<>();
        output3.put("greeting", "Hello, Alice!");
        output3.put("ageNextYear", 31);
        String fp3 = fingerprint(input3, output3);
        System.out.println("Test 3 (object): " + fp3);

        // Test 4: array output (List)
        List<Object> input4 = Arrays.asList("a", "b", "c");
        List<Object> output4 = Arrays.asList("A", "B", "C");
        String fp4 = fingerprint(input4, output4);
        System.out.println("Test 4 (array): " + fp4);

        // Test 5: null input
        String fp5 = fingerprint(null, "default");
        System.out.println("Test 5 (null input): " + fp5);

        // Test 6: NaN output (sentinel)
        String fp6 = fingerprint(0.0, Double.NaN);
        System.out.println("Test 6 (NaN output): " + fp6);
    }
}

File: scripts/scala/regret_fingerprint.scala
// regret_fingerprint.scala — Scala port of scripts/fingerprint.js
//
// Cross-stack fingerprint parity contract:
//   fingerprint(input, output)
//   = sha256(stableStringify(input) + "|" + stableStringify(output))
//   → hex → BigInt → base36 → first 7 chars
//
// MUST produce byte-identical output to scripts/fingerprint.js (JS),
// scripts/fingerprint.py (Python), scripts/fingerprint_php.php (PHP),
// scripts/fingerprint_bash.sh (Bash).
//
// The module is self-contained: no external deps. We hand-roll stable JSON
// serialization so we can match JS byte-for-byte (e.g. NaN sentinel, integer
// vs float, key sorting).

package regrets

import java.math.BigInteger
import java.security.MessageDigest

/** A tiny JSON value tree — sufficient for our serialize needs. */
enum Json:
  case Null
  case Bool(b: Boolean)
  case IntNum(n: Long)
  case DoubleNum(d: Double)
  case Str(s: String)
  case Arr(items: Array[Json])
  case Obj(entries: Array[(String, Json)])

object Json:
  /** Parse a JSON string into our tree. */
  def parse(s: String): Json = JsonParser.parse(s)

  /** Build a JSON tree from a Scala Any (best-effort). */
  def from(a: Any): Json = a match
    case null                   => Null
    case j: Json                => j
    case b: Boolean             => Bool(b)
    case b: java.lang.Boolean   => Bool(b.booleanValue())
    case b: Byte                => IntNum(b.toLong)
    case s: Short               => IntNum(s.toLong)
    case i: Int                 => IntNum(i.toLong)
    case l: Long                => IntNum(l)
    case f: Float               => DoubleNum(f.toDouble)
    case d: Double              => DoubleNum(d)
    case s: String              => Str(s)
    case s: CharSequence        => Str(s.toString)
    case arr: Array[?]          => Arr(arr.map(from).asInstanceOf[Array[Json]])
    case seq: Seq[?]            => Arr(seq.map(from).toArray)
    case m: Map[?, ?]           =>
      val entries = m.toArray.map { case (k, v) => (k.toString, from(v)) }
      Obj(entries)
    case p: Product             =>
      val names = p.productElementNames.toArray
      val vals  = p.productIterator.toArray
      Obj(names.zip(vals).map { case (n, v) => (n, from(v)) })
    case other                  => Str(other.toString)
/** A minimal, strict JSON parser producing `Json` values. */
object JsonParser:
  import Json._

  def parse(s: String): Json =
    val p = new Parser(s)
    val v = p.skipWs().parseValue()
    p.skipWs()
    if p.pos < p.input.length then
      throw new RuntimeException(s"Trailing garbage at position ${p.pos}")
    v

  private class Parser(val input: String):
    var pos: Int = 0

    def skipWs(): this.type =
      while pos < input.length && (input.charAt(pos) match
          case ' ' | '\t' | '\n' | '\r' => true
          case _ => false
      ) do pos += 1
      this

    def parseValue(): Json =
      skipWs()
      if pos >= input.length then throw new RuntimeException("Unexpected EOF")
      input.charAt(pos) match
        case '"'  => Str(parseString())
        case '{'  => parseObject()
        case '['  => parseArray()
        case 't'  => parseLiteral("true", Bool(true))
        case 'f'  => parseLiteral("false", Bool(false))
        case 'n'  => parseLiteral("null", Null)
        case '-' | '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' =>
          parseNumber()
        case c => throw new RuntimeException(s"Unexpected char '$c' at $pos")

    def parseLiteral(lit: String, v: Json): Json =
      if input.regionMatches(pos, lit, 0, lit.length) then
        pos += lit.length
        v
      else throw new RuntimeException(s"Expected '$lit' at $pos")

    def parseString(): String =
      // Assumes current char is '"'
      pos += 1
      val sb = new StringBuilder
      var escape = false
      var done = false
      while pos < input.length && !done do
        val c = input.charAt(pos); pos += 1
        if escape then
          escape = false
          c match
            case '"'  => sb.append('"')
            case '\\' => sb.append('\\')
            case '/'  => sb.append('/')
            case 'b'  => sb.append('\b')
            case 'f'  => sb.append('\f')
            case 'n'  => sb.append('\n')
            case 'r'  => sb.append('\r')
            case 't'  => sb.append('\t')
            case 'u'  =>
              if pos + 4 > input.length then throw new RuntimeException("Bad \\u escape")
              val hex = input.substring(pos, pos + 4)
              pos += 4
              sb.append(Integer.parseInt(hex, 16).toChar)
            case _ => throw new RuntimeException(s"Bad escape '\\$c'")
        else c match
          case '"'  => done = true
          case '\\' => escape = true
          case _    => sb.append(c)
      if !done then throw new RuntimeException("Unterminated string")
      sb.toString

    def parseNumber(): Json =
      val start = pos
      if input.charAt(pos) == '-' then pos += 1
      while pos < input.length && input.charAt(pos).isDigit do pos += 1
      var isDouble = false
      if pos < input.length && input.charAt(pos) == '.' then
        isDouble = true
        pos += 1
        while pos < input.length && input.charAt(pos).isDigit do pos += 1
      if pos < input.length && (input.charAt(pos) == 'e' || input.charAt(pos) == 'E') then
        isDouble = true
        pos += 1
        if pos < input.length && (input.charAt(pos) == '+' || input.charAt(pos) == '-') then pos += 1
        while pos < input.length && input.charAt(pos).isDigit do pos += 1
      val text = input.substring(start, pos)
      if isDouble then DoubleNum(text.toDouble)
      else
        try IntNum(text.toLong)
        catch case _: Exception => DoubleNum(text.toDouble)

    def parseArray(): Json =
      pos += 1 // skip [
      skipWs()
      val items = scala.collection.mutable.ArrayBuffer[Json]()
      if pos < input.length && input.charAt(pos) == ']' then
        pos += 1
        return Arr(items.toArray)
      var done = false
      while !done do
        items += parseValue()
        skipWs()
        if pos >= input.length then throw new RuntimeException("Unterminated array")
        input.charAt(pos) match
          case ',' => pos += 1; skipWs()
          case ']' => pos += 1; done = true
          case _   => throw new RuntimeException(s"Expected ',' or ']' at $pos")
      Arr(items.toArray)

    def parseObject(): Json =
      pos += 1 // skip {
      skipWs()
      val entries = scala.collection.mutable.ArrayBuffer[(String, Json)]()
      if pos < input.length && input.charAt(pos) == '}' then
        pos += 1
        return Obj(entries.toArray)
      var done = false
      while !done do
        skipWs()
        if pos >= input.length || input.charAt(pos) != '"' then
          throw new RuntimeException(s"Expected string key at $pos")
        val key = parseString()
        skipWs()
        if pos >= input.length || input.charAt(pos) != ':' then
          throw new RuntimeException(s"Expected ':' at $pos")
        pos += 1
        val v = parseValue()
        entries += (key -> v)
        skipWs()
        if pos >= input.length then throw new RuntimeException("Unterminated object")
        input.charAt(pos) match
          case ',' => pos += 1
          case '}' => pos += 1; done = true
          case _   => throw new RuntimeException(s"Expected ',' or '}' at $pos")
      Obj(entries.toArray)

// ─── Stable serialization (parity with JS stableStringify) ────────────────

object StableStringify:
  /** Serialize a Json value into a stable, byte-deterministic JSON string. */
  def apply(v: Json): String =
    val sb = new StringBuilder
    render(v, sb)
    sb.toString

  private def render(v: Json, sb: StringBuilder): Unit = v match
    case Json.Null          => sb.append("null")
    case Json.Bool(true)    => sb.append("true")
    case Json.Bool(false)   => sb.append("false")
    case Json.IntNum(n)     => sb.append(n.toString)
    case Json.DoubleNum(d)  =>
      if java.lang.Double.isNaN(d)        then sb.append("\"__nan__\"")
      else if d == Double.PositiveInfinity then sb.append("\"__infinity__\"")
      else if d == Double.NegativeInfinity then sb.append("\"__neg_infinity__\"")
      else sb.append(formatDouble(d))
    case Json.Str(s)        => renderString(s, sb)
    case Json.Arr(items)    =>
      sb.append('[')
      var first = true
      for it <- items do
        if !first then sb.append(',')
        first = false
        render(it, sb)
      sb.append(']')
    case Json.Obj(entries)  =>
      val sorted = entries.sortBy(_._1)
      sb.append('{')
      var first = true
      for (k, v) <- sorted do
        if !first then sb.append(',')
        first = false
        renderString(k, sb)
        sb.append(':')
        render(v, sb)
      sb.append('}')

  private def renderString(s: String, sb: StringBuilder): Unit =
    sb.append('"')
    var i = 0
    while i < s.length do
      val c = s.charAt(i)
      c match
        case '"'  => sb.append("\\\"")
        case '\\' => sb.append("\\\\")
        case '\b' => sb.append("\\b")
        case '\f' => sb.append("\\f")
        case '\n' => sb.append("\\n")
        case '\r' => sb.append("\\r")
        case '\t' => sb.append("\\t")
        case _ =>
          if c < 0x20 then
            sb.append("\\u")
            sb.append(f"${c.toInt}%04x")
          else
            sb.append(c)
      i += 1
    sb.append('"')

  /** Format a Double to match JSON.stringify output. */
  private def formatDouble(d: Double): String =
    if (d == math.floor(d)) && !java.lang.Double.isInfinite(d) && math.abs(d) < 1e21 then
      new java.math.BigDecimal(d.toLong).toPlainString
    else
      val s = d.toString
      // Scala: 1.0E7 / 1.0E-7  →  JS: 1e+7 / 1e-7
      s.replace("E", "e+").replace("e+-", "e-")

// ─── Fingerprint (parity with JS fingerprint.js) ─────────────────────────

object Fingerprint:
  /** Compute the 7-char base36 fingerprint for an (input, output) pair. */
  def apply(input: Json, output: Json): String =
    val combined = StableStringify(input) + "|" + StableStringify(output)
    val md = MessageDigest.getInstance("SHA-256")
    val bytes = md.digest(combined.getBytes("UTF-8"))
    val hex = bytes.map(b => f"${b & 0xff}%02x").mkString
    val big = new BigInteger(hex, 16)
    val b36 = big.toString(36)
    if b36.length >= 7 then b36.substring(0, 7) else b36
